lib/os/fdtable.c

```diff
@@ -167,27 +170,66 @@ int z_alloc_fd(void *obj, const struct fd_op_vtable *vtable)
 
 #ifdef CONFIG_POSIX_API
 
-ssize_t read(int fd, void *buf, size_t sz)
+ssize_t z_impl_sys_read(int fd, void *buf, size_t sz)
 {
 	if (_check_fd(fd) < 0) {
 		return -1;
 	}
 
 	return fdtable[fd].vtable->read(fdtable[fd].obj, buf, sz);
 }
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_sys_read(int fd, void *buf, size_t sz)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(buf, sz)) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	return z_impl_sys_read(fd, buf, sz);
+}
+#include <syscalls/sys_read_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/* Normal C function wrapping a corresponding syscall. Required to ensure
+ * classic C linkage.
+ */
+ssize_t read(int fd, void *buf, size_t sz)
+{
+	return sys_read(fd, buf, sz);
+}
 FUNC_ALIAS(read, _read, ssize_t);
 
-ssize_t write(int fd, const void *buf, size_t sz)
+ssize_t z_impl_sys_write(int fd, const void *buf, size_t sz)
 {
 	if (_check_fd(fd) < 0) {
 		return -1;
 	}
 
 	return fdtable[fd].vtable->write(fdtable[fd].obj, buf, sz);
 }
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_sys_write(int fd, const void *buf, size_t sz)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, sz));
```

Comment:
Don't we need to make a copy of the buf data here?

Reply:
No, why?

Reply:
Just looking the existing calls in sockets.c, but indeed we do not copy the data buffer there either.

Reply:
We need to copy only stuff which may lead to security issues. E.g., if some memory block contains pointers, we need to copy those pointers (not necessarily the entire memory block!), then validate the copies, then let the kernel use copies for dereferencing.

(Yeah, yeah, someone looking for side-channel (e.g., timing) attacks hardening may copy more stuff, e.g. control-dependent data. Good luck with that to whoever will do that.)

Reply:
The contents of buf are really application dependent, and generally are not used by the kernel, unless it is a kernel type of socket / block / char device.

Validating the application has access to the memory should be sufficient.